qupid/matching.py

```diff
@@ -0,0 +1,79 @@
+import collections
+import random
+
+from networkx.algorithms.bipartite import sets as bipartite_sets
+
+INFINITY = float("inf")
+
+
+def hopcroft_karp_matching(G, top_nodes=None):
+    """Returns a maximum cardinality matching of the bipartite graph G.
+
+    NOTE: This function is modified from the NetworkX implementation:
+        https://tinyurl.com/akcfy6r4
+
+    We have modified the DFS algorithm to move to a *random* neighbor from all
```

Comment:
Would you get the same effect by shuffling cases and controls prior to running the default hopcroft karp?  That seems much easier than maintaining this random neighbor search implementation separate from the networkx implementation.

Reply:
From my understanding of the way NX works, shuffling the dictionary prior to creating the Graph would not address our concerns. I guess a possible alternative is to shuffle within the Graph but there's no easy way to do this so it would require some hacking and also be done at each iteration probably.